From b60e4088827ab04e0f4d3f24c65c265293024869 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Wed, 28 Feb 2024 16:47:34 +0100
Subject: [PATCH 1063/1141] media: dcmipp: pixelproc: fix {H/V}RATIO of
 downsize

Depending on the source and destination size it may happen with
previous processing that the last line of the image is dropped,
leaving a frame for 1 line less than requested.  In order to avoid
having a downsize ratio too high and leading to this issue, avoid
performing processing on original/destination size - 1.

Change-Id: I7bde7569dc09892b640dbea39f6dc42774e630ca
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/363602
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Philippe CORNU <philippe.cornu@foss.st.com>
Tested-by: Jerome BESNARD <jerome.besnard@st.com>
Domain-Review: Philippe CORNU <philippe.cornu@foss.st.com>
---
 .../platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c     | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
index 2977b6c06fc3..d120f69f08e2 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
@@ -705,12 +705,12 @@ dcmipp_pixelproc_set_downscale(struct dcmipp_pixelproc_device *pixelproc)
 	}
 
 	/* Compute downsize factor */
-	hratio = (h_post_dec - 1) * DCMIPP_PIXELPROC_HVRATIO_CONS /
-		 (pixelproc->compose.width - 1);
+	hratio = h_post_dec * DCMIPP_PIXELPROC_HVRATIO_CONS /
+		 pixelproc->compose.width;
 	if (hratio > DCMIPP_PIXELPROC_HVRATIO_MAX)
 		hratio = DCMIPP_PIXELPROC_HVRATIO_MAX;
-	vratio = (v_post_dec - 1) * DCMIPP_PIXELPROC_HVRATIO_CONS /
-		 (pixelproc->compose.height - 1);
+	vratio = v_post_dec * DCMIPP_PIXELPROC_HVRATIO_CONS /
+		 pixelproc->compose.height;
 	if (vratio > DCMIPP_PIXELPROC_HVRATIO_MAX)
 		vratio = DCMIPP_PIXELPROC_HVRATIO_MAX;
 	hdiv = (DCMIPP_PIXELPROC_HVDIV_CONS * pixelproc->compose.width) /
-- 
2.39.2

